### PR TITLE
Fix flaky unit tests

### LINKED ISSUE #5313: Several unit tests are flaky and need to get fixed
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:

/kind bug
> /kind enhancement

**What happened**:

With PR we introduced some changes to our testing on CI that a) avoids caching of test results and b) tests multiple times. Now a couple of PRs already have suffered by the unit test lane `pull-kubevirt-unit-test` blocking merges due to flaky unit tests.

The test suites discovered that were flaky are (please add to this list so we have it complete)

*  TestVirtOperator
* TestVirtWrap
* TestMigrationProxy
* TestCache
* TestHeartbeat
* TestComponents

Source PR histories:
* [#5308](https://prow.apps.ovirt.org/pr-history/?org=kubevirt&repo=kubevirt&pr=5308)
* [#5300](https://prow.apps.ovirt.org/pr-history/?org=kubevirt&repo=kubevirt&pr=5300)
* [#5311](https://prow.apps.ovirt.org/pr-history/?org=kubevirt&repo=kubevirt&pr=5311)

Possible candidates (couldn't verify flakiness yet):
* [#5022](https://prow.apps.ovirt.org/pr-history/?org=kubevirt&repo=kubevirt&pr=5022)
 
See #5308, #5312

/cc @rmohr @fgimenez 

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by dhiller**
**What this PR does / why we need it**:
This PR fixes flaky unit tests by eliminating data races that occur during using go `-race` flag. Based on initial changes from rmohrs PR they isolate test data and synchronize go routines to avoid concurrency issues. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5313

**Special notes for your reviewer**:

This is a spinoff from https://github.com/kubevirt/kubevirt/pull/5262 in order to get the test changes in and help fixing test flake issues.

**Release note**:

```release-note
Fix flaky unit tests
```

### CODE REVIEW

File: pkg/virt-operator/kubevirt_test.go [hunk 1]
```diff
@@ -2555,6 +2555,44 @@ var _ = Describe("KubeVirt Operator", func() {
 		}, 15)
 	})
 
+	Context("when the monitor namespace does not exist", func() {
```
Comment (rmohr) [inline pkg/virt-operator/kubevirt_test.go:2573]: this test looks like it was added in this commit and not moved. Can it be that the original one was not deleted?

**Review by rmohr — COMMENTED** [review]
One question. I think this PR can basically go in to allow enabling code coverage reports and race checking for our code.

It outlines some areas which we can tackle:

 * some tests just take way too long
 * I consider the operator race fixes as a "valid hack". It makes sense to use a struct here ,but I also have the feeling that we don't have clear control of the shutdown process of the operator.

I also wonder why you seem to succeed in running tests in parallel, because I do not even manage to run the unit tests outside of the build container without #4793.

**Comment by EdDev** [comment]
General not: This change does not seem to be an user-facing one, therefore the release-note is redundant.

**Comment by EdDev** [comment]
This PR seems to be a continuation of https://github.com/kubevirt/kubevirt/pull/4887 , right?
@dhiller , could you please feedback on the comments given there?

File: pkg/virt-operator/kubevirt_test.go (hunk 1 above)
Comment (dhiller) [reply pkg/virt-operator/kubevirt_test.go:2573]: Thanks for the catch, will have a look!

Comment (dhiller) [reply pkg/virt-operator/kubevirt_test.go:2530]: OK, I removed the duplicate test.

**Comment by dhiller** [comment]
> This PR seems to be a continuation of #4887 , right?
> @dhiller , could you please feedback on the comments given there?

I think the answers @rmohr gave were pretty exhaustive. Can you elaborate where you want me to comment further? BTW as that PR is closed, I'm unsure if it is of value if we do that work there?

**Review by rmohr — COMMENTED** [review]
You may want  to squash the commits a little, like doing the test case move in one commit. The content looks good enough. :+1:

**Comment by dhiller** [comment]
@rmohr sqashed the commits together, PTAL, thanks!

**Pull request merged**